dlt/pipeline/helpers.py

```diff
@@ -32,4 +36,100 @@ def _retry_load(ex: Exception) -> bool:
+        self.schema_tables = self.schema.tables
+
+        drop_tables = []
+        for tbl in self.schema_tables.values():
+            if tbl.get('resource') in resources:
+                drop_tables += get_child_tables(self.schema_tables, tbl['name'])
+        self.drop_tables = list(reversed(drop_tables))
+
+        # # List resource state keys to drop
+        # drop_resources = set(r for r in (t.get('resource') for t in self.drop_tables) if r)
+        self.drop_state_paths: List[JSONPath] = [jsonpath_parse(p) for p in state_paths or []]
+        self.drop_resource_state_keys = resources
+
+    def drop_destination_tables(self) -> None:
+        with self.pipeline._get_destination_client(self.schema) as client:
+            client.drop_tables(*[tbl['name'] for tbl in self.drop_tables])
+
+    def delete_pipeline_tables(self) -> None:
+        for tbl in self.drop_tables:
+            del self.schema_tables[tbl['name']]
+        self.schema.bump_version()
+
+    def _list_state_paths(self, source_state: Dict[str, Any]) -> Iterator[str]:
+        for path in self.drop_state_paths:
+            matches = path.find(source_state)
+            for match in matches:
+                yield str(match.full_path)
+
+    def _drop_state_paths(self, source_state: Dict[str, Any]) -> None:
+        for path in self.drop_state_paths:
+            path.filter(lambda _: True, source_state)
+
+    def drop_state_keys(self) -> None:
+        state: TSourceState
+        with self.pipeline.managed_state(extract_state=True, extract_unchanged=True) as state:  # type: ignore[assignment]
+            source_states = state.get("sources")
+            if not source_states:
+                return
+            # TODO: Should only drop from source state matching the schema. Not possible atm
+            for source_state in source_states.values():
+                for p in self.drop_state_paths:  # Remove keys by JSON path
+                    p.filter(lambda _: True, source_state)
+                resources = source_state.get('resources')
+                if not resources:
+                    return
+                for key in self.drop_resource_state_keys:
+                    resources.pop(key, None)
+
+    def info(self) -> Any:
+        return dict(
+            drop_tables=[tbl['name'] for tbl in self.drop_tables],
+        )
+
+    def __call__(self) -> None:
+        if self.pipeline.has_pending_data:  # Raise when there are pending extracted/load files to prevent conflicts
+            raise PipelineHasPendingDataException(self.pipeline.pipeline_name, self.pipeline.pipelines_dir)
+        self.delete_pipeline_tables()
+        self.drop_destination_tables()
+        self.drop_state_keys()
+        self.pipeline.schemas.save_schema(self.schema)
+        # Send updated state to destination
+        self.pipeline.normalize()
+        self.pipeline._get_load_storage().delete_completed_package
```

Comment:
that does nothing?

Reply:
Yeah just clutter leftover. Was trying something.